File: packages/providers/src/openai-provider.ts
import OpenAI from 'openai';
import type {
  Message,
  LlmResponse,
  StreamEvent,
  ToolDescription,
  ToolCall,
  ProviderConfig,
} from '@cli-agent/core';
import { ProviderError } from '@cli-agent/core';
import { BaseProvider } from './base-provider.js';
import { extractToken } from './auth/auth-resolver.js';

export class OpenAIProvider extends BaseProvider {
  readonly providerId = 'openai';
  private readonly client: OpenAI;
  private readonly model: string;
  private readonly maxTokens: number;
  private readonly temperature: number;

  constructor(config: ProviderConfig) {
    super('openai-provider');
    const apiKey = extractToken(config.auth);
    this.client = new OpenAI({ apiKey, baseURL: config.baseUrl });
    this.model = config.model;
    this.maxTokens = config.maxTokens;
    this.temperature = config.temperature;
  }

  async chat(
    messages: readonly Message[],
    tools?: readonly ToolDescription[]
  ): Promise<LlmResponse> {
    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        max_tokens: this.maxTokens,
        temperature: this.temperature,
        messages: this.toOpenAIMessages(messages),
        tools: tools && tools.length > 0 ? this.toOpenAITools(tools) : undefined,
      });

      return this.parseResponse(response);
    } catch (error) {
      throw new ProviderError(
        `OpenAI API error: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  async *stream(
    messages: readonly Message[],
    tools?: readonly ToolDescription[]
  ): AsyncIterable<StreamEvent> {
    try {
      const stream = await this.client.chat.completions.create({
        model: this.model,
        max_tokens: this.maxTokens,
        temperature: this.temperature,
        messages: this.toOpenAIMessages(messages),
        tools: tools && tools.length > 0 ? this.toOpenAITools(tools) : undefined,
        stream: true,
      });

      let content = '';
      const toolCalls: ToolCall[] = [];
      let inputTokens = 0;
      let outputTokens = 0;

      for await (const chunk of stream) {
        const delta = chunk.choices[0]?.delta;
        if (!delta) continue;

        if (delta.content) {
          content += delta.content;
          yield { type: 'text_delta', content: delta.content };
        }

        if (delta.tool_calls) {
          for (const tc of delta.tool_calls) {
            if (tc.id) {
              toolCalls.push({
                id: tc.id,
                name: tc.function?.name ?? '',
                arguments: tc.function?.arguments ?? '',
              });
              yield {
                type: 'tool_call_start',
                toolCall: { id: tc.id, name: tc.function?.name },
              };
            } else if (tc.function?.arguments) {
              const last = toolCalls[toolCalls.length - 1];
              if (last) {
                toolCalls[toolCalls.length - 1] = {
                  ...last,
                  arguments: last.arguments + tc.function.arguments,
                };
              }
              yield { type: 'tool_call_delta', content: tc.function.arguments };
            }
          }
        }

        if (chunk.usage) {
          inputTokens = chunk.usage.prompt_tokens;
          outputTokens = chunk.usage.completion_tokens;
        }

        if (chunk.choices[0]?.finish_reason) {
          const stopReason =
            chunk.choices[0].finish_reason === 'tool_calls'
              ? 'tool_use' as const
              : chunk.choices[0].finish_reason === 'length'
                ? 'max_tokens' as const
                : 'end_turn' as const;

          yield {
            type: 'done',
            response: {
              content,
              stopReason,
              toolCalls,
              usage: { inputTokens, outputTokens },
            },
          };
        }
      }
    } catch (error) {
      throw new ProviderError(
        `OpenAI stream error: ${error instanceof Error ? error.message : String(error)}`,
        error instanceof Error ? error : undefined
      );
    }
  }

  private toOpenAIMessages(
    messages: readonly Message[]
  ): OpenAI.ChatCompletionMessageParam[] {
    const result: OpenAI.ChatCompletionMessageParam[] = [];

    for (const msg of messages) {
      if (msg.role === 'system') {
        result.push({ role: 'system', content: msg.content });
      } else if (msg.toolResults && msg.toolResults.length > 0) {
        for (const tr of msg.toolResults) {
          result.push({
            role: 'tool',
            tool_call_id: tr.toolCallId,
            content: tr.content,
          });
        }
      } else if (msg.role === 'assistant' && msg.toolCalls && msg.toolCalls.length > 0) {
        result.push({
          role: 'assistant',
          content: msg.content || null,
          tool_calls: msg.toolCalls.map((tc) => ({
            id: tc.id,
            type: 'function' as const,
            function: { name: tc.name, arguments: tc.arguments },
          })),
        });
      } else if (msg.role === 'assistant') {
        result.push({ role: 'assistant', content: msg.content });
      } else {
        result.push({ role: 'user', content: msg.content });
      }
    }

    return result;
  }

  private toOpenAITools(
    tools: readonly ToolDescription[]
  ): OpenAI.ChatCompletionTool[] {
    return tools.map((tool) => ({
      type: 'function' as const,
      function: {
        name: tool.name,
        description: tool.description,
        parameters: {
          type: 'object',
          properties: Object.fromEntries(
            tool.parameters.map((p) => [
              p.name,
              { type: p.type, description: p.description },
            ])
          ),
          required: tool.parameters
            .filter((p) => p.required)
            .map((p) => p.name),
        },
      },
    }));
  }

  private parseResponse(
    response: OpenAI.ChatCompletion
  ): LlmResponse {
    const choice = response.choices[0];
    if (!choice) {
      throw new ProviderError('No choices in OpenAI response');
    }

    const toolCalls: ToolCall[] = (choice.message.tool_calls ?? []).map((tc) => ({
      id: tc.id,
      name: tc.function.name,
      arguments: tc.function.arguments,
    }));

    const stopReason =
      choice.finish_reason === 'tool_calls'
        ? 'tool_use' as const
        : choice.finish_reason === 'length'
          ? 'max_tokens' as const
          : 'end_turn' as const;

    return {
      content: choice.message.content ?? '',
      stopReason,
      toolCalls,
      usage: {
        inputTokens: response.usage?.prompt_tokens ?? 0,
        outputTokens: response.usage?.completion_tokens ?? 0,
      },
    };
  }
}

File: sdk/python/chameleon/providers/openai_provider.py
"""
OpenAI LLM provider — Python port of @cli-agent/providers OpenAIProvider.
"""

from __future__ import annotations

import json
import logging
from typing import Any, AsyncIterator

from openai import AsyncOpenAI

from ..types import (
    ILlmProvider,
    LlmResponse,
    StreamEvent,
    ToolCall,
    ToolDescription,
    TokenUsage,
)

logger = logging.getLogger(__name__)


class OpenAIProvider(ILlmProvider):
    """Async OpenAI chat completions provider."""

    def __init__(
        self,
        api_key: str,
        model: str = "gpt-4o",
        max_tokens: int = 4096,
        temperature: float = 0.7,
        base_url: str | None = None,
    ) -> None:
        self._client = AsyncOpenAI(api_key=api_key, base_url=base_url)
        self._model = model
        self._max_tokens = max_tokens
        self._temperature = temperature

    @property
    def provider_id(self) -> str:
        return "openai"

    async def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[ToolDescription] | None = None,
    ) -> LlmResponse:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "temperature": self._temperature,
            "messages": self._to_openai_messages(messages),
        }

        openai_tools = self._to_openai_tools(tools) if tools else None
        if openai_tools:
            kwargs["tools"] = openai_tools

        response = await self._client.chat.completions.create(**kwargs)

        choice = response.choices[0]
        tool_calls = [
            ToolCall(
                id=tc.id,
                name=tc.function.name,
                arguments=tc.function.arguments,
            )
            for tc in (choice.message.tool_calls or [])
        ]

        stop_reason = (
            "tool_use"
            if choice.finish_reason == "tool_calls"
            else "max_tokens"
            if choice.finish_reason == "length"
            else "end_turn"
        )

        return LlmResponse(
            content=choice.message.content or "",
            stop_reason=stop_reason,
            tool_calls=tool_calls,
            usage=TokenUsage(
                input_tokens=response.usage.prompt_tokens if response.usage else 0,
                output_tokens=response.usage.completion_tokens if response.usage else 0,
            ),
        )

    async def stream(
        self,
        messages: list[dict[str, Any]],
        tools: list[ToolDescription] | None = None,
    ) -> AsyncIterator[StreamEvent]:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "temperature": self._temperature,
            "messages": self._to_openai_messages(messages),
            "stream": True,
        }

        openai_tools = self._to_openai_tools(tools) if tools else None
        if openai_tools:
            kwargs["tools"] = openai_tools

        stream = await self._client.chat.completions.create(**kwargs)

        content = ""
        tool_calls: list[ToolCall] = []

        async for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta is None:
                continue

            if delta.content:
                content += delta.content
                yield StreamEvent(type="text_delta", content=delta.content)

            if delta.tool_calls:
                for tc in delta.tool_calls:
                    if tc.id:
                        tool_calls.append(
                            ToolCall(
                                id=tc.id,
                                name=tc.function.name if tc.function else "",
                                arguments=tc.function.arguments if tc.function else "",
                            )
                        )
                        yield StreamEvent(type="tool_call_start", content=tc.function.name if tc.function else "")
                    elif tc.function and tc.function.arguments:
                        if tool_calls:
                            last = tool_calls[-1]
                            tool_calls[-1] = ToolCall(
                                id=last.id,
                                name=last.name,
                                arguments=last.arguments + tc.function.arguments,
                            )
                        yield StreamEvent(type="tool_call_delta", content=tc.function.arguments)

            finish = chunk.choices[0].finish_reason if chunk.choices else None
            if finish:
                stop_reason = (
                    "tool_use" if finish == "tool_calls"
                    else "max_tokens" if finish == "length"
                    else "end_turn"
                )
                yield StreamEvent(
                    type="done",
                    response=LlmResponse(
                        content=content,
                        stop_reason=stop_reason,
                        tool_calls=tool_calls,
                    ),
                )

    # ----- helpers -----

    @staticmethod
    def _to_openai_messages(
        messages: list[dict[str, Any]],
    ) -> list[dict[str, Any]]:
        result: list[dict[str, Any]] = []
        for msg in messages:
            role = msg.get("role", "user")
            if role == "tool":
                result.append({
                    "role": "tool",
                    "tool_call_id": msg.get("tool_call_id", ""),
                    "content": msg.get("content", ""),
                })
            elif role == "assistant" and "tool_calls" in msg:
                result.append({
                    "role": "assistant",
                    "content": msg.get("content") or None,
                    "tool_calls": [
                        {
                            "id": tc["id"],
                            "type": "function",
                            "function": {
                                "name": tc["name"],
                                "arguments": tc["arguments"],
                            },
                        }
                        for tc in msg["tool_calls"]
                    ],
                })
            else:
                result.append({"role": role, "content": msg.get("content", "")})
        return result

    @staticmethod
    def _to_openai_tools(
        tools: list[ToolDescription],
    ) -> list[dict[str, Any]]:
        return [
            {
                "type": "function",
                "function": {
                    "name": t.name,
                    "description": t.description,
                    "parameters": {
                        "type": "object",
                        "properties": {
                            p.name: {"type": p.type, "description": p.description}
                            for p in t.parameters
                        },
                        "required": [p.name for p in t.parameters if p.required],
                    },
                },
            }
            for t in tools
        ]
